token/program-2022/src/extension/token_metadata/processor.rs

```diff
@@ -1,18 +1,88 @@
 //! Token-metadata processor
 
 use {
-    solana_program::{account_info::AccountInfo, entrypoint::ProgramResult, msg, pubkey::Pubkey},
-    spl_token_metadata_interface::instruction::{
-        Emit, Initialize, RemoveKey, TokenMetadataInstruction, UpdateAuthority, UpdateField,
+    crate::{
+        check_program_account,
+        error::TokenError,
+        extension::{
+            alloc_and_serialize, metadata_pointer::MetadataPointer, BaseStateWithExtensions,
+            StateWithExtensions,
+        },
+        state::Mint,
+    },
+    solana_program::{
+        account_info::{next_account_info, AccountInfo},
+        entrypoint::ProgramResult,
+        msg,
+        program_error::ProgramError,
+        program_option::COption,
+        pubkey::Pubkey,
+    },
+    spl_token_metadata_interface::{
+        error::TokenMetadataError,
+        instruction::{
+            Emit, Initialize, RemoveKey, TokenMetadataInstruction, UpdateAuthority, UpdateField,
+        },
+        state::{OptionalNonZeroPubkey, TokenMetadata},
     },
 };
 
 /// Processes a [Initialize](enum.TokenMetadataInstruction.html) instruction.
 pub fn process_initialize(
     _program_id: &Pubkey,
-    _accounts: &[AccountInfo],
-    _data: Initialize,
+    accounts: &[AccountInfo],
+    data: Initialize,
 ) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let metadata_info = next_account_info(account_info_iter)?;
+    let update_authority_info = next_account_info(account_info_iter)?;
+    let mint_info = next_account_info(account_info_iter)?;
+    let mint_authority_info = next_account_info(account_info_iter)?;
+
+    // check that the mint and metadata accounts are the same, since the metadata
+    // extension should only describe itself
+    if metadata_info.key != mint_info.key {
+        msg!("Metadata for a mint must be initialized in the mint itself.");
+        return Err(TokenError::MintMismatch.into());
+    }
```

Comment:
Sorry, I'm not sure I understand the question.  Are you suggesting to move the metadata pointer process to another place?

Also, note that there are no PDAs owned by token-2022 at all, everything is in the mint, including the pointer and the metadata.

Reply:
I’m only raising this point because I thought it was previously discussed that Token2022 _could_ implement the metadata interface in two ways - both within the mint and separated state.

Perhaps I misunderstood and the idea was merely to support only the pointer if your metadata is elsewhere.

If you've no intention of supporting separate PDAs and the program will only handle Mint accounts then we're good to go

Reply:
Ah I see what you mean now. Since token accounts don't have a discriminator, except when extended, it's really tricky to add a new account type, so the concept is to keep everything just inside the mint.